src/Development/IDE/Core/Compile.hs

```diff
@@ -682,39 +683,55 @@ loadInterface
   -> (Maybe LinkableType -> m ([FileDiagnostic], Maybe HiFileResult)) -- ^ Action to regenerate an interface
   -> m ([FileDiagnostic], Maybe HiFileResult)
 loadInterface session ms sourceMod linkableNeeded regen = do
-    res <- liftIO $ checkOldIface session ms sourceMod Nothing
+    let session' = session {hsc_dflags = foldl' gopt_set (ms_hspp_opts ms) [Opt_IgnoreOptimChanges,Opt_IgnoreHpcChanges]}
+
+    -- We need to read in an old iface, because checkOldIface won't if the source is changed
+    -- (since we set HscNothing)
+    mold_iface <- liftIO $ initIfaceCheck (text "loadInterface") session' $
+      readIface (ms_mod ms) (msHiFilePath ms)
```

Comment:
I don't think it matters. `checkOldIface` attempts to read the old iface in most scenarios, but doesn't do so if the target is `HscNothing`, and just returns a `Nothing`, forcing recompilation. If the target it set to anything else, it will try to read the old iface itself (if we didn't pass one to it).

By reading the iface ourselves, we bypass all logic in `checkOldIface` to read the iface, making it have the desired behavior regardless of `hscTarget`.

Reply:
It doesn't matter what `hscTarget` is set to. Since we read the iface here, `checkOldIface` will not attempt to read the iface again.

Reply:
Fair enough!